### PR TITLE
Sets sysctls from pod config annotations

### PR DESCRIPTION
**Pull request by mikebrow**
Tested via e2e node.. results attached.

Signed-off-by: Mike Brown <brownwm@us.ibm.com>

### CODE REVIEW

**Comment by mikebrow** [comment]
mike@mike-VirtualBox:~/go/src/k8s.io/kubernetes$ sudo make test-e2e-node FOCUS=Sysctls RUNTIME=remote CONTAINER_RUNTIME_ENDPOINT=/var/run/cri-containerd.sock
[sudo] password for mike: 
+++ [0808 17:50:28] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
    k8s.io/kubernetes/vendor/github.com/jteeuwen/go-bindata/go-bindata
+++ [0808 17:50:28] Generating bindata:
    test/e2e/generated/gobindata_util.go
~/go/src/k8s.io/kubernetes ~/go/src/k8s.io/kubernetes/test/e2e/generated
~/go/src/k8s.io/kubernetes/test/e2e/generated
+++ [0808 17:50:29] Building go targets for linux/amd64:
    vendor/github.com/onsi/ginkgo/ginkgo
Creating artifacts directory at /tmp/_artifacts/170808T175124
Test artifacts will be written to /tmp/_artifacts/170808T175124
Updating sudo credentials
+++ [0808 17:51:28] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
    k8s.io/kubernetes/vendor/github.com/jteeuwen/go-bindata/go-bindata
+++ [0808 17:51:28] Generating bindata:
    test/e2e/generated/gobindata_util.go
~/go/src/k8s.io/kubernetes ~/go/src/k8s.io/kubernetes/test/e2e/generated
~/go/src/k8s.io/kubernetes/test/e2e/generated
+++ [0808 17:51:30] Building go targets for linux/amd64:
    cmd/kubelet
    test/e2e_node/e2e_node.test
    vendor/github.com/onsi/ginkgo/ginkgo
Running Suite: E2eNode Suite
============================
Random Seed: 1502232812 - Will randomize all specs
Will run 225 specs

Running in parallel across 8 nodes

I0808 17:53:33.356706   13964 validators.go:44] Validating os...
OS: Linux
I0808 17:53:33.370943   13964 validators.go:44] Validating kernel...
I0808 17:53:33.372696   13964 kernel_validator.go:77] Validating kernel version
KERNEL_VERSION: 4.4.0-89-generic
I0808 17:53:33.372817   13964 kernel_validator.go:92] Validating kernel config
CONFIG_NAMESPACES: enabled
CONFIG_NET_NS: enabled
CONFIG_PID_NS: enabled
CONFIG_IPC_NS: enabled
CONFIG_UTS_NS: enabled
CONFIG_CGROUPS: enabled
CONFIG_CGROUP_CPUACCT: enabled
CONFIG_CGROUP_DEVICE: enabled
CONFIG_CGROUP_FREEZER: enabled
CONFIG_CGROUP_SCHED: enabled
CONFIG_CPUSETS: enabled
CONFIG_MEMCG: enabled
CONFIG_INET: enabled
CONFIG_EXT4_FS: enabled
CONFIG_PROC_FS: enabled
CONFIG_NETFILTER_XT_TARGET_REDIRECT: enabled (as module)
CONFIG_NETFILTER_XT_MATCH_COMMENT: enabled (as module)
CONFIG_OVERLAY_FS: enabled (as module)
CONFIG_AUFS_FS: enabled (as module)
CONFIG_BLK_DEV_DM: enabled
I0808 17:53:33.411591   13964 validators.go:44] Validating cgroups...
CGROUPS_CPU: enabled
CGROUPS_CPUACCT: enabled
CGROUPS_CPUSET: enabled
CGROUPS_DEVICES: enabled
CGROUPS_FREEZER: enabled
CGROUPS_MEMORY: enabled
I0808 17:53:33.411711   13964 validators.go:44] Validating package...
PASS
I0808 17:53:33.413257   13903 e2e_node_suite_test.go:135] Pre-pulling images so that they are cached for the tests.
I0808 17:53:33.413269   13903 remote_image.go:40] Connecting to image service /var/run/cri-containerd.sock
W0808 17:53:33.413294   13903 util_linux.go:75] Using "/var/run/cri-containerd.sock" as endpoint is deprecated, please consider using full url format "unix:///var/run/cri-containerd.sock".
I0808 17:53:33.413654   13903 image_list.go:136] Pre-pulling images with CRI [gcr.io/google-containers/stress:v1 gcr.io/google_containers/busybox:1.24 gcr.io/google_containers/busybox@sha256:4bdd623e848417d96127e16037743f0cd8b528c026e9175e22a84f639eca58ff gcr.io/google_containers/eptest:0.1 gcr.io/google_containers/hostexec:1.2 gcr.io/google_containers/liveness:e2e gcr.io/google_containers/mounttest-user:0.5 gcr.io/google_containers/mounttest:0.8 gcr.io/google_containers/netexec:1.4 gcr.io/google_containers/netexec:1.5 gcr.io/google_containers/netexec:1.7 gcr.io/google_containers/nginx-slim:0.7 gcr.io/google_containers/node-problem-detector:v0.4.0 gcr.io/google_containers/pause-amd64:3.0 gcr.io/google_containers/serve_hostname:v1.4 gcr.io/google_containers/test-webserver:e2e gcr.io/google_containers/volume-gluster:0.2 gcr.io/google_containers/volume-nfs:0.8 google/cadvisor:latest]
I0808 17:53:53.071617   13903 kubelet.go:93] Starting kubelet
I0808 17:53:53.088450   13903 server.go:147] Starting server "kubelet" with command "/usr/bin/systemd-run --unit=kubelet-924491080.service --slice=runtime.slice --remain-after-exit /home/mike/go/src/k8s.io/kubernetes/_output/local/go/bin/kubelet --kubelet-cgroups=/kubelet.slice --cgroup-root=/ --api-servers http://localhost:8080 --address 0.0.0.0 --port 10250 --read-only-port 10255 --volume-stats-agg-period 10s --allow-privileged true --serialize-image-pulls false --pod-manifest-path /home/mike/go/src/k8s.io/kubernetes/_output/local/go/bin/pod-manifest947325206 --file-check-frequency 10s --pod-cidr 10.180.0.0/24 --eviction-pressure-transition-period 30s --feature-gates  --eviction-hard memory.available<250Mi,nodefs.available<10%%,nodefs.inodesFree<5%% --eviction-minimum-reclaim nodefs.available=5%%,nodefs.inodesFree=5%% --v 4 --logtostderr --network-plugin=kubenet --cni-bin-dir /home/mike/go/src/k8s.io/kubernetes/_output/local/go/bin/cni/bin --cni-conf-dir /home/mike/go/src/k8s.io/kubernetes/_output/local/go/bin/cni/net.d --hostname-override mike-VirtualBox --container-runtime-endpoint=/var/run/cri-containerd.sock --container-runtime=remote --network-plugin= --network-plugin-dir="
I0808 17:53:53.088526   13903 server.go:102] Running readiness check for service "kubelet"
I0808 17:53:53.088586   13903 server.go:175] Output file for server "kubelet": /tmp/_artifacts/170808T175124/kubelet.log
I0808 17:53:53.108813   13903 server.go:217] Running health check for service "kubelet"
I0808 17:53:53.108840   13903 server.go:102] Running readiness check for service "kubelet"
I0808 17:53:56.091286   13903 server.go:147] Starting server "services" with command "/home/mike/go/src/k8s.io/kubernetes/_output/local/go/bin/e2e_node.test --run-services-mode --test.timeout=24h --ginkgo.seed=1502232812 --ginkgo.focus=Sysctls --ginkgo.skip=\\[Flaky\\]|\\[Slow\\]|\\[Serial\\] --ginkgo.parallel.node=1 --ginkgo.parallel.total=8 --ginkgo.parallel.streamhost=http://127.0.0.1:33878 --ginkgo.parallel.synchost=http://127.0.0.1:33878 --ginkgo.slowSpecThreshold=5.00000 --container-runtime=remote --container-runtime-endpoint=/var/run/cri-containerd.sock --image-service-endpoint= --alsologtostderr --v 4 --report-dir=/tmp/_artifacts/170808T175124 --node-name mike-VirtualBox --kubelet-flags=--container-runtime-endpoint=/var/run/cri-containerd.sock --kubelet-flags=--container-runtime=remote --kubelet-flags=--network-plugin= --network-plugin-dir="
I0808 17:53:56.091427   13903 server.go:102] Running readiness check for service "services"
I0808 17:53:56.091478   13903 server.go:175] Output file for server "services": /tmp/_artifacts/170808T175124/services.log
I0808 17:53:56.095039   13903 server.go:206] Waiting for server "services" start command to complete
I0808 17:53:56.113885   13903 server.go:228] Initial health check passed for service "kubelet"
I0808 17:54:02.112350   13903 e2e_node_suite_test.go:150] Node services started.  Running tests...
I0808 17:54:02.112359   13903 e2e_node_suite_test.go:155] Wait for the node to be ready

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[BeforeEach] [k8s.io] Sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:130
STEP: Creating a kubernetes client
STEP: Building a namespace api object
Aug  8 17:54:03.396: INFO: Skipping waiting for service account
[BeforeEach] [k8s.io] Sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/sysctl.go:58
[It] should reject invalid sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/sysctl.go:182
STEP: Creating a pod with one valid and two invalid sysctls
[AfterEach] [k8s.io] Sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:131
Aug  8 17:54:03.401: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "e2e-tests-sysctl-d6wg9" for this suite.
Aug  8 17:54:11.437: INFO: namespace: e2e-tests-sysctl-d6wg9, resource: bindings, ignored listing per whitelist
Aug  8 17:54:11.447: INFO: namespace e2e-tests-sysctl-d6wg9 deletion completed in 8.043246009s


• [SLOW TEST:8.061 seconds]
[k8s.io] Sysctls
/home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:619
  should reject invalid sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/sysctl.go:182
------------------------------
[BeforeEach] [k8s.io] Sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:130
STEP: Creating a kubernetes client
STEP: Building a namespace api object
Aug  8 17:54:03.261: INFO: Skipping waiting for service account
[BeforeEach] [k8s.io] Sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/sysctl.go:58
[It] should support unsafe sysctls which are actually whitelisted
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/sysctl.go:140
STEP: Creating a pod with the kernel.shm_rmid_forced sysctl
STEP: Watching for error events or started pod
STEP: Waiting for pod completion
STEP: Checking that the pod succeeded
STEP: Getting logs from the pod
STEP: Checking that the sysctl is actually updated
[AfterEach] [k8s.io] Sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:131
Aug  8 17:54:15.287: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "e2e-tests-sysctl-ww21b" for this suite.
Aug  8 17:54:27.315: INFO: namespace: e2e-tests-sysctl-ww21b, resource: bindings, ignored listing per whitelist
Aug  8 17:54:27.320: INFO: namespace e2e-tests-sysctl-ww21b deletion completed in 12.032215778s


• [SLOW TEST:24.099 seconds]
[k8s.io] Sysctls
/home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:619
  should support unsafe sysctls which are actually whitelisted
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/sysctl.go:140
------------------------------
[BeforeEach] [k8s.io] Sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:130
STEP: Creating a kubernetes client
STEP: Building a namespace api object
Aug  8 17:54:03.367: INFO: Skipping waiting for service account
[BeforeEach] [k8s.io] Sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/sysctl.go:58
[It] should not launch unsafe, but not explicitly enabled sysctls on the node
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/sysctl.go:209
STEP: Creating a pod with a greylisted, but not whitelisted sysctl on the node
STEP: Watching for error events or started pod
STEP: Checking that the pod was rejected
[AfterEach] [k8s.io] Sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:131
Aug  8 17:54:15.381: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "e2e-tests-sysctl-fskhq" for this suite.
Aug  8 17:54:31.406: INFO: namespace: e2e-tests-sysctl-fskhq, resource: bindings, ignored listing per whitelist
Aug  8 17:54:31.420: INFO: namespace e2e-tests-sysctl-fskhq deletion completed in 16.038085963s


• [SLOW TEST:28.064 seconds]
[k8s.io] Sysctls
/home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:619
  should not launch unsafe, but not explicitly enabled sysctls on the node
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/sysctl.go:209
------------------------------
[BeforeEach] [k8s.io] Sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:130
STEP: Creating a kubernetes client
STEP: Building a namespace api object
Aug  8 17:54:03.427: INFO: Skipping waiting for service account
[BeforeEach] [k8s.io] Sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/sysctl.go:58
[It] should support sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/sysctl.go:99
STEP: Creating a pod with the kernel.shm_rmid_forced sysctl
STEP: Watching for error events or started pod
STEP: Waiting for pod completion
STEP: Checking that the pod succeeded
STEP: Getting logs from the pod
STEP: Checking that the sysctl is actually updated
[AfterEach] [k8s.io] Sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:131
Aug  8 17:54:15.447: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "e2e-tests-sysctl-gpmzr" for this suite.
Aug  8 17:54:33.479: INFO: namespace: e2e-tests-sysctl-gpmzr, resource: bindings, ignored listing per whitelist
Aug  8 17:54:33.481: INFO: namespace e2e-tests-sysctl-gpmzr deletion completed in 18.032353446s


• [SLOW TEST:30.058 seconds]
[k8s.io] Sysctls
/home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:619
  should support sysctls
  /home/mike/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/sysctl.go:99
------------------------------
I0808 17:54:33.539538   13903 e2e_node_suite_test.go:171] Stopping node services...
I0808 17:54:33.539582   13903 server.go:303] Kill server "services"
I0808 17:54:33.539603   13903 server.go:340] Killing process 14013 (services) with -TERM
I0808 17:54:33.646644   13903 server.go:303] Kill server "kubelet"
I0808 17:54:33.673851   13903 services.go:156] Fetching log files...
I0808 17:54:33.673912   13903 services.go:165] Get log file "docker.log" with journalctl command [-u docker].
I0808 17:54:33.688603   13903 services.go:165] Get log file "cloud-init.log" with journalctl command [-u cloud*].
E0808 17:54:33.692851   13903 services.go:168] failed to get "cloud-init.log" from journald: Failed to add filter for units: No data available
, exit status 1
I0808 17:54:33.693076   13903 services.go:165] Get log file "kubelet.log" with journalctl command [-u kubelet-924491080.service].
I0808 17:54:33.835290   13903 services.go:165] Get log file "kern.log" with journalctl command [-k].
I0808 17:54:33.873568   13903 e2e_node_suite_test.go:176] Tests Finished


Ran 4 of 225 Specs in 60.907 seconds
SUCCESS! -- 4 Passed | 0 Failed | 0 Pending | 221 Skipped 

Ginkgo ran 1 suite in 1m1.697931208s
Test Suite Passed

**Comment by mikebrow** [comment]
Had to modify https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/sysctl/runtime.go#L75-L81 to  
```
	return &runtimeAdmitHandler{
		result: lifecycle.PodAdmitResult{
			Admit: true,
		},
	}, nil
```
To convince e2e to run the tests... will open an issue.

**Comment by Random-Liu** [comment]
PR is sent on the Kubernetes side https://github.com/kubernetes/kubernetes/pull/50390.

**Comment by Random-Liu** [comment]
@mikebrow Sysctls are part of CRI api now. We should use that directly instead of getting out from annotation.

This will make the code cleaner, too.

**Comment by mikebrow** [comment]
hmm.. looks like it failed the "unsafe" sysctls

File: pkg/server/sandbox_run.go [hunk 1]
```diff
@@ -306,7 +312,11 @@ func (c *criContainerdService) generateSandboxContainerSpec(id string, config *r
 
 	// TODO(random-liu): [P1] Set privileged.
 
-	// TODO(random-liu): [P2] Set sysctl from annotations.
+	// add sysctls
```
Comment (Random-Liu) [inline pkg/server/sandbox_run.go]: nit: s/add/Add.

File: pkg/server/sandbox_run.go [hunk 2]
```diff
@@ -268,7 +268,13 @@ func (c *criContainerdService) generateSandboxContainerSpec(id string, config *r
 	// Set hostname.
 	g.SetHostname(config.GetHostname())
 
-	// TODO(random-liu): [P2] Consider whether to add labels and annotations to the container.
+	// get annotations
+	annotations := config.GetAnnotations()
```
Comment (Random-Liu) [inline pkg/server/sandbox_run.go]: Don't think we need this now. :)

**Comment by Random-Liu** [comment]
The code become much smaller...yeah. :)
@mikebrow apparmor and seccomp profile are both in CRI now, that will also be straightforward. Just reference at the implementation in dockershim.

**Comment by mikebrow** [comment]
"Just reference at the implementation in dockershim." roger that

**Comment by Random-Liu** [comment]
LGTM, will merge after test passes.

**Pull request merged**